### DISCUSSION TITLE
Provide a working example how to use renovate with Google Artifact Registry

### DISCUSSION THREAD

**Discussion by jagoral** [renovatebot/renovate#24249] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

In my project I use some packages which are published in Google Artifact Registry. I want to authenticate the Mend Renovate to be able to check for the dependency updates for these packages. Google Artifact Registry generates a short-lived access token, so to make it work, I've tried to pass a user credentials, same as described [here](https://github.com/renovatebot/renovate/blob/main/docs/usage/docker.md#google-container-registry--google-artifact-registry), however none of the `hostRules` configuration worked, and Artifact Registry returned 403. 

One of the configuration which I've tried

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", ":semanticCommitTypeAll(chore)"],
  "timezone": "Europe/Warsaw",
  "branchPrefix": "deps/",
  "baseBranches": ["main"],
  "rangeStrategy": "bump",
  "npm": {
    "commitMessageTopic": "{{prettyDepType}} {{depName}}",
    "commitMessagePrefix": "chore(deps): :arrow_up: "
  },
  "schedule": ["after 2am and before 3am"],
  "hostRules": [
    {
      "matchHost": "europe-west1-npm.pkg.dev",
      "username": "_json_key_base64",
      "encrypted": {
        "password": "<base_64_service_account_encrypted_with_app.renovatebot.com/encrypt"
      }
    }
  ]
}
```

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by Shegox** [comment]
You should probably be able to follow the instructions from [Set up authentication for npm - Configuring password authentication](https://cloud.google.com/artifact-registry/docs/nodejs/authentication#auth-password) and add the resulting `.npmrc` encrypted to Renovate. 

Interestingly enough the generated npmrc seems like it contains the username `_json_key_base64` and as password a twice(!) base64 encoded service account json: 
```
@SCOPE:registry=https://europe-npm.pkg.dev/my-project/renovate-test/
//europe-npm.pkg.dev/my-project/renovate-test/:always-auth=true

# Insert the following snippet into your user .npmrc

//europe-npm.pkg.dev/my-project/renovate-test/:_password="<twice base64 encoded service account json>"
//europe-npm.pkg.dev/my-project/renovate-test/:username=_json_key_base64
//europe-npm.pkg.dev/my-project/renovate-test/:email=not.valid@email.com
```

~~Could you try encoding the service account twice with base64 `cat service-account.json | base64 | base64 ` and then encrypt it?~~

--- 

Edit: I just tested it myself and the key only needs to be base64 encoded once. Afterwards adding this hostRule did the trick.
```js
    {
      matchHost: "europe-npm.pkg.dev",
      username: "_json_key_base64",
      password: "<base64 encoded key>"
    }
```

and my .npmrc looks like this: 
```npmrc
registry=https://europe-npm.pkg.dev/my-project/my-repository/
//europe-npm.pkg.dev/my-project/my-repository/:always-auth=true
```

So your configuration looks right, can you check the logs what error message is thrown when authenticating?

--- 

Related discussion: #23124

**Comment by jagoral** [comment]
Thanks, @Shegox, I will check it, and let you know if it helped! :)

**Comment by jagoral** [comment]
This is what I get, when I try to run renovate through `renovate` CLI
```
DEBUG: GET https://europe-west1-npm.pkg.dev/my-project/npm/@scope%2Fpackage = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=372) (repository=myrepo, baseBranch=xxx)
DEBUG: Failed to look up npm package @scope/package (repository=myrepo, baseBranch=xxx, packageFile=xxx/package.json, dependency=@scope/package)
```

my `renovate.json` (same as in the first message, I've tried with and without `encrypted`)

my `.npmrc`
```
@scope:registry=https://europe-west1-npm.pkg.dev/my-project/npm/
//europe-west1-npm.pkg.dev/my-project/npm/:always-auth=true
```

**Reply by Shegox** [reply]
And just to be sure, the GCP service account has access to the artifact registry, right?

**Reply by jagoral** [reply]
right

**Reply by Shegox** [reply]
It took me a bit of debugging, but I think I found the problem: The host rule for NPM must include `"hostType": "npm"`. Can you try that?

You can find my full test repository (using the official Mend Renovate) here: https://github.com/Shegox/renovate-gar-test

--- 

If that is really the case, I think we should improve our documentation in that regard. Especially the section about private modules: https://docs.renovatebot.com/getting-started/private-packages/#npm

**Reply by ghost** [reply]
(context: I am Tomasz's coworker)

Hah I was copying the _password generated from the .npmrc and trying to use it not knowing it was double base64'd

I couldn't get it to work using the tips from above though. The service account works fine for me, but not for Renovate - ie. I can use the .npmrc myself to run `npm view @scope/myprivatepkg`, but when I base64 it once and give it to renovate to run locally, it doesn't work and returns 403.

```
  ...,
  "hostRules": [
    {
      "matchHost": "europe-west1-npm.pkg.dev",
      "hostType": "npm",
      "username": "_json_key_base64",
      "password": "result of 'cat gcp_service_account.json | base64'"
    }
  ],
  "npmrc": "@myscope:registry=https://europe-west1-npm.pkg.dev/myrepo/npm/\n//europe-west1-npm.pkg.dev/myrepo/npm/:always-auth=true"
}
```

LOG_LEVEL debug snippet:
```
DEBUG: GET https://europe-west1-npm.pkg.dev/myrepo/npm/@myscope%2Fredacted = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=530) (repository=redacted, baseBranch=redacted)
DEBUG: Failed to look up npm package @myscope/redacted (repository=vuestorefront/redacted, baseBranch=redacted, packageFile=packages/redacted/package.json, dependency=@myscope/redacted)
```

Is there anything else I can try to provide more diagnostic info?

**Reply by Shegox** [reply]
@sethidden to be honest I'm out of all ideas now.... 😅 

One last thing you can try is running Renovate with `LOG_LEVEL=TRACE` to get very detailed logs and check them if the token is set correctly in the header or see if the response contains any more information.

One more thing, can you try adding `\\n` to your npmrc instead of `\n`. In JSON backslashes should be must be escaped. For me it worked with one as well, but thats the last straw I could imagine being wrong with your config.

**Reply by viceice** [reply]
don't do the base64 before passing to renovate host rules, renovate will do the base64 before writing to npmrc

**Reply by Shegox** [reply]
@viceice not in this case, as GCP actually expects as password the full base64 encoded service-account.json. So you need to manually base64 encode it once.

**Reply by michaelKurowski** [reply]
They way how we solved it was making GCP output an NPM-friendly token (`token_format: access_token` if you use `google-github-actions/auth@v1` github action). We then append `npmrc` option in the renovate config with the following string `//url-to-the-registry/:_authToken=here-we-place-the-token`. Docs mention that you can put ${NPM_TOKEN} instead of a hardcoded token, but it didn't work for us, so we hardcode it.
